File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<int[]> queue = new LinkedList<>(); 
        int total_weight = 0;
        int index = 0;

        while (!queue.isEmpty() || index < truck_weights.length) {
            answer++; // 시간 흐름
            
            // 다리에서 나갈 트럭 제거
            if (!queue.isEmpty()) {
                if (answer >= queue.peek()[1] + bridge_length) {
                    int w = queue.poll()[0];
                    total_weight -= w;
                }
            }

            // 트럭 진입
            if (index < truck_weights.length &&
                total_weight + truck_weights[index] <= weight &&
                queue.size() < bridge_length) {
                total_weight += truck_weights[index];
                queue.offer(new int[]{truck_weights[index], answer});
                index++;
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0;
    let bridge = Array(bridge_length).fill(0); // 다리 위 상태
    let totalWeight = 0;

    while (truck_weights.length > 0 || totalWeight > 0) {
        time++;

        // 1초 경과 → 다리 앞부분 트럭 제거
        totalWeight -= bridge.shift();

        // 다음 트럭이 올라갈 수 있으면
        if (truck_weights.length > 0 && totalWeight + truck_weights[0] <= weight) {
            let truck = truck_weights.shift();
            bridge.push(truck);
            totalWeight += truck;
        } else {
            // 아니면 0 (빈 공간)
            bridge.push(0);
        }
    }

    return time;
}
